openedx/core/djangoapps/content_libraries/models.py

```diff
@@ -127,3 +202,248 @@ def save(self, *args, **kwargs):  # lint-amnesty, pylint: disable=arguments-diff
+                    "to.")
+    )
+
+    client_id = models.CharField(
+        max_length=255,
+        verbose_name=_('client identifier'),
+        help_text=_("The LTI client identifier generated by the platform.")
+    )
+
+    subject_id = models.CharField(
+        max_length=255,
+        verbose_name=_('subject identifier'),
+        help_text=_('Identifies the entity that initiated the launch request, '
+                    'commonly a user.')
+    )
+
+    created_at = models.DateTimeField(
+        auto_now_add=True
+    )
+
+    class Meta:
+        unique_together = ['platform_id', 'client_id', 'subject_id']
+
+    @property
+    def subject_url(self):
+        """
+        An local URL that is known to uniquely identify this profile.
+
+        We take advantage of the fact that platform id is required to be an URL
+        and append paths with the reamaining keys to it.
+        """
+        return '/'.join([
+            self.platform_id.rstrip('/'),
+            self.client_id,
+            self.subject_id
+        ])
+
+    def save(self, *args, **kwds):
+        """
+        Get or create an edx user on save.
+        """
+        if not self.user:
+            uid = uuid.uuid5(uuid.NAMESPACE_URL, self.subject_url)
+            username = f'urn:openedx:content_libraries:username:{uid}'
+            email = f'{uid}@{ContentLibrariesConfig.name}'
+            with transaction.atomic():
+                if self.user is None:
+                    self.user, created = User.objects.get_or_create(
+                        username=username,
+                        defaults={'email': email})
+                    if created:
+                        # LTI users can only auth throught LTI launches.
+                        self.user.set_unusable_password()
+                    self.user.save()
+                super().save(*args, **kwds)
+
+    def __str__(self):
+        return self.subject_id
+
+
+class LtiGradedResourceManager(models.Manager):
+    """
+    A custom manager for the graded resources model.
+    """
+
+    def upsert_from_ags_launch(self, user, block, resource_endpoint, resource_link):
+        """
+        Update or create a graded resource at AGS launch.
+        """
+        resource_id = resource_link['id']
+        resource_title = resource_link.get('title') or None
+        lineitem = resource_endpoint['lineitem']
+        lti_profile = user.contentlibraries_lti_profile
+        try:
+            resource = self.get(
+                profile=lti_profile,
+                usage_key=block.scope_ids.usage_id
+            )
+        except self.model.DoesNotExist:
+            resource = self.model(
+                profile=lti_profile,
+                usage_key=block.scope_ids.usage_id,
+            )
+        resource.resource_title = resource_title
+        resource.resource_id = resource_id
+        resource.ags_lineitem = lineitem
+        resource.save()
```

Comment:
Nit: Django has [`update_or_create`](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#update-or-create) which you can use to code this sort of upsert more declaratively and with fewer lines.